File: app/src/main/java/com/mudrahome/MHLMS/Activities/LeadListActivity.java
package com.mudrahome.MHLMS.Activities;

import android.annotation.SuppressLint;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBarDrawerToggle;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.widget.Toolbar;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.FragmentTransaction;
import androidx.viewpager.widget.ViewPager;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.google.android.material.tabs.TabLayout;
import com.mudrahome.MHLMS.Adapters.LeadListPagerAdapter;
import com.mudrahome.MHLMS.Firebase.Authentication;
import com.mudrahome.MHLMS.Firebase.Firestore;
import com.mudrahome.MHLMS.Fragments.ChangePasswordFragment;
import com.mudrahome.MHLMS.Fragments.LeadListFragment;
import com.mudrahome.MHLMS.Interfaces.FirestoreInterfaces;
import com.mudrahome.MHLMS.Interfaces.OnPasswordChange;
//import com.mudrahome.MHLMS.Managers.LeadManager;
import com.mudrahome.MHLMS.Managers.ProfileManager;
import com.mudrahome.MHLMS.Models.UserDetails;
import com.mudrahome.MHLMS.R;
import com.mudrahome.MHLMS.SharedPreferences.UserDataSharedPreference;


public class LeadListActivity extends BaseActivity implements NavigationView.OnNavigationItemSelectedListener {

    private ProfileManager profileManager;
    private Toolbar toolbar;
    private Firestore firestore;
    private TabLayout tabLayout;
    DrawerLayout drawerLayout;
    NavigationView navigationView;
    private ActionBarDrawerToggle actionBarDrawerToggle;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.drawer_layout_leadlist);

        toolbar = findViewById(R.id.toolbarLeadList);
        toolbar.inflateMenu(R.menu.lead_list_menu);
        setSupportActionBar(toolbar);

        tabLayout = findViewById(R.id.tabLayout);
        drawerLayout = findViewById(R.id.drawer_layout);
        navigationView = findViewById(R.id.navigationdrawer_dashboard);

        profileManager = new ProfileManager();
        firestore = new Firestore();

        actionBarDrawerToggle = new ActionBarDrawerToggle(
                this, drawerLayout, R.string.navigation_drawer_open, R.string.navigation_drawer_close);
        drawerLayout.addDrawerListener(actionBarDrawerToggle);
        actionBarDrawerToggle.syncState();
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        navigationView.setNavigationItemSelectedListener(this);

        firestore.getUsers(new FirestoreInterfaces.OnGetUserDetails() {
            @Override
            public void onSuccess(UserDetails userDetails) {
                profileManager.setCurrentUserDetails(userDetails);

                if (profileManager.getCurrentUserType().contains(getString(R.string.admin)) &&
                        profileManager.getCurrentUserType().contains(getString(R.string.salesman))) {
                    openViewPager(R.string.admin_and_salesman);
                } else if (profileManager.getCurrentUserType().contains(getString(R.string.telecaller))) {
                    openFragment(R.string.telecaller);
                } else if (profileManager.getCurrentUserType().contains(getString(R.string.admin))) {
                    openFragment(R.string.admin);

                } else {
                    openFragment(R.string.salesman);
                }
            }

            @Override
            public void fail() {

            }
        }, profileManager.getuId());
    }

    private void openFragment(int userType) {
        FragmentTransaction ft = getSupportFragmentManager().beginTransaction();
        ft.replace(R.id.frame_layout, new LeadListFragment(userType));
        ft.commit();
    }

    private void openViewPager(int userType) {
        tabLayout.setVisibility(View.VISIBLE);
        openFragment(R.string.admin);
        final ViewPager vpPager = findViewById(R.id.pager);
        LeadListPagerAdapter adapterViewPager = new LeadListPagerAdapter(getSupportFragmentManager(), 2);
        vpPager.setAdapter(adapterViewPager);

        vpPager.addOnPageChangeListener(new TabLayout.TabLayoutOnPageChangeListener(tabLayout));

        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                vpPager.setCurrentItem(tab.getPosition());
                if (tab.getPosition() == 0) {
                    openFragment(R.string.admin);
                } else {
                    openFragment(R.string.salesman);
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {

            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {

            }
        });
    }

    @Override
    public void onBackPressed() {
        //super.onBackPressed();
        Intent intent = new Intent();
        intent.putExtra("loggedIn", profileManager.checkUserExist());
        setResult(101, intent);
        finish();
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        drawerLayout.closeDrawers();

        switch (menuItem.getItemId()) {
            case R.id.profileDetails:
                startActivity(new Intent(LeadListActivity.this, ProfileDetailsActivity.class));
                break;

            case R.id.change_password:
                if (isNetworkConnected()) {
                    showPasswordFragment();
                } else {
                    showToastMessage(R.string.no_internet);
                }
                break;

            case R.id.logout:
                if (isNetworkConnected()) {
                    showLogOutWarning();

                } else {
                    showToastMessage(R.string.no_internet);
                }
                break;
        }
        return true;
    }

    private void showLogOutWarning() {
        AlertDialog.Builder build = new AlertDialog.Builder(LeadListActivity.this);
        build.setMessage("Are you sure you want to logout?")
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int u) {
                        profileManager.signOut();
                        showToastMessage(R.string.logged_out);
//                        kghv k
                        //SharedPreferences.Editor editor = sharedPreferences.edit();
                        //editor.clear();
                        onBackPressed();
                    }
                }).setNegativeButton("No", (dialogInterface, u) -> {

        });
        androidx.appcompat.app.AlertDialog alert = build.create();
        alert.show();
    }

    private void showPasswordFragment() {
        final UserDataSharedPreference userDataSharedPreference =
                new UserDataSharedPreference(LeadListActivity.this);

        ChangePasswordFragment.newInstance((oldPassword, newPassword) -> {

            showProgressDialog("Please wait...", LeadListActivity.this);
            Authentication authentication = new Authentication(LeadListActivity.this);
            authentication.UpdatePassword(
                    oldPassword,
                    newPassword,
                    userDataSharedPreference.getUserEmail(),
                    new OnPasswordChange() {
                        @Override
                        public void onSucess(String result) {

                            hideKeyboard(LeadListActivity.this);
                            Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
                            dismissProgressDialog();

                        }
                    });
        }).show(getSupportFragmentManager(), "changepassword");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.lead_list_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle item selection
        switch (item.getItemId()) {
            case R.id.notification:
                if (isNetworkConnected()) {
                    Intent intent = new Intent(LeadListActivity.this, NotificationActivity.class);
                    startActivity(intent);
                } else
                    showToastMessage(R.string.no_internet);
                break;

            case android.R.id.home:
                drawerLayout.openDrawer(GravityCompat.START);
                break;
        }
        return super.onOptionsItemSelected(item);
    }


}

File: app/src/main/java/com/mudrahome/mhlms/activities/LeadListActivity.kt
package com.mudrahome.mhlms.activities

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import com.google.android.material.navigation.NavigationView
import com.mudrahome.mhlms.R
import com.mudrahome.mhlms.adapters.LeadListPagerAdapter
import com.mudrahome.mhlms.databinding.ActivityLeadListBinding
import com.mudrahome.mhlms.firebase.Authentication
import com.mudrahome.mhlms.firebase.Firestore
import com.mudrahome.mhlms.fragments.ChangePasswordFragment
import com.mudrahome.mhlms.fragments.LeadListFragment
import com.mudrahome.mhlms.interfaces.FirestoreInterfaces
import com.mudrahome.mhlms.managers.ProfileManager
import com.mudrahome.mhlms.model.UserDetails
import com.mudrahome.mhlms.sharedPreferences.ProfileSP

class LeadListActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener {

    private var profileManager: ProfileManager? = null
    private var firestore: Firestore? = null
    private var toggle: ActionBarDrawerToggle? = null
    private var userType: Int? = null
    private var binding: ActivityLeadListBinding? = null
    private var profileSP: ProfileSP? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_lead_list)

        setSupportActionBar(binding!!.toolbar)

        profileManager = ProfileManager()
        firestore = Firestore()
        profileSP = ProfileSP(this)

        toggle = ActionBarDrawerToggle(
            this,
            binding?.drawerLayout,
            binding?.toolbar,
            R.string.navigation_drawer_open,
            R.string.navigation_drawer_close
        )

        binding?.drawerLayout?.addDrawerListener(toggle!!)
        toggle!!.syncState()
        binding!!.navView.setNavigationItemSelectedListener(this)

        showProgressDialog("Loading..", this)

        firestore!!.getUsers(object : FirestoreInterfaces.OnGetUserDetails {
            override fun onSuccess(userDetails: UserDetails) {

                if (progress.isShowing) dismissProgressDialog()

                if (userDetails.userType!!.contains(getString(R.string.admin)) &&
                    userDetails.userType!!.contains(getString(R.string.salesman))
                ) {
                    userType = R.string.admin_and_salesman
                    openViewPager(userType!!)

                } else if (userDetails.userType!!.contains(getString(R.string.telecaller)) &&
                    userDetails.userType!!.contains(getString(R.string.teleassigner))
                ) {
                    userType = R.string.telecaller_and_teleassigner
                    openViewPager(userType!!)

                } else if (userDetails.userType!!.contains(getString(R.string.telecaller))) {
                    userType = R.string.telecaller
                    openFragment(userType!!)

                } else if (userDetails.userType!!.contains(getString(R.string.admin))) {
                    userType = R.string.admin
                    openFragment(userType!!)

                } else if (userDetails.userType!!.contains(getString(R.string.teleassigner))) {
                    userType = R.string.teleassigner
                    openFragment(userType!!)

                } else {
                    userType = R.string.salesman
                    openFragment(userType!!)
                }
            }

            override fun fail() {

            }
        }, profileManager!!.getuId())
    }

    private fun openFragment(userType: Int) {
        Log.d("dsfvdfv", userType.toString())

        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.frame_layout, LeadListFragment(userType))
        transaction.commit()
    }

    private fun openViewPager(userType: Int) {
        binding!!.tabLayout.visibility = View.VISIBLE

        val pagerAdapter = LeadListPagerAdapter(supportFragmentManager)

        val tab1title: String?
        val tab2title: String?

        val type1: Int?
        val type2: Int?

        if (userType == R.string.telecaller_and_teleassigner) {
            tab1title = "Caller"
            tab2title = "Assign"
            type1 = R.string.telecaller
            type2 = R.string.teleassigner

        } else {
            tab1title = "Admin"
            tab2title = "Sales"
            type1 = R.string.admin
            type2 = R.string.salesman
        }
        pagerAdapter.addFragment(LeadListFragment(type1), tab1title)
        pagerAdapter.addFragment(LeadListFragment(type2), tab2title)

        binding!!.pager.adapter = pagerAdapter
        binding!!.tabLayout.setupWithViewPager(binding!!.pager)
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        binding!!.drawerLayout.closeDrawers()

        when (menuItem.itemId) {
            R.id.profileDetails -> startActivity(
                Intent(
                    this@LeadListActivity,
                    ProfileDetailsActivity::class.java
                )
            )

            R.id.change_password -> if (isNetworkConnected) {
                showPasswordFragment()
            } else {
                showToastMessage(R.string.no_internet)
            }

            R.id.logout -> if (isNetworkConnected) {
                showLogOutWarning()

            } else {
                showToastMessage(R.string.no_internet)
            }
        }
        return true
    }

    private fun showLogOutWarning() {
        val build = AlertDialog.Builder(this@LeadListActivity)
        build.setMessage("Are you sure you want to logout?")
            .setPositiveButton("Yes") { _, _ ->
                profileManager!!.signOut()
                profileSP!!.clearSharePreference()                                      // Clear data from cache
                showToastMessage(R.string.logged_out)
                val intent = Intent(this@LeadListActivity, LoginActivity::class.java)
                intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP
                startActivity(intent)
                finish()
            }.setNegativeButton("No") { _, _ ->

            }
        val alert = build.create()
        alert.show()
    }

    private fun showPasswordFragment() {
        val userDataSharedPreference =
            ProfileSP(this@LeadListActivity)

        ChangePasswordFragment.newInstance { oldPassword, newPassword ->

            showProgressDialog("Please wait...", this@LeadListActivity)
            val authentication = Authentication(this@LeadListActivity)
            authentication.UpdatePassword(

                oldPassword,
                newPassword,
                userDataSharedPreference.userEmail
            ) { result ->

                hideKeyboard(this@LeadListActivity)
                Toast.makeText(applicationContext, result, Toast.LENGTH_SHORT).show()
                dismissProgressDialog()

            }
        }.show(supportFragmentManager, "changepassword")
    }
}